File: tools/xgmtool/src/samplebank.c
#include <stdlib.h>
#include <memory.h>

#include "../inc/samplebank.h"
#include "../inc/vgmcom.h"
#include "../inc/xgmtool.h"


SampleBank* SampleBank_create(VGMCommand* command)
{
    if (!VGMCommand_isDataBlock(command))
    {
        printf("Error: incorrect sample data declaration at %6X !\n", command->offset);
        return NULL;
    }

    SampleBank* result;

    result = malloc(sizeof(SampleBank));

    result->data = command->data;
    result->offset = command->offset;

    // id
    result->id = VGMCommand_getDataBankId(command);
    // len
    result->len = VGMCommand_getDataBlockLen(command);

    if (verbose)
        printf("Initial bank sample added [%6X-%6X]   rate: %d Hz\n", 0, result->len - 1, 0);

    // consider the whole bank as a single sample by default
    result->samples = createElement(Sample_create(0, 0, result->len, 0));

    return result;
}

void SampleBank_addBlock(SampleBank* bank, VGMCommand* command)
{
    if (!VGMCommand_isDataBlock(command))
        return;

    // concat data block
    int newLen = bank->len + VGMCommand_getDataBlockLen(command);
    unsigned char* newData = malloc(newLen + 7);

    memcpy(&newData[0], &bank->data[bank->offset], bank->len + 7);
    memcpy(&newData[bank->len + 7], &command->data[command->offset + 7], VGMCommand_getDataBlockLen(command));
    // adjust len
    setInt(newData, 0 + 3, newLen);

    if (verbose)
        printf("Initial block sample added [%6X-%6X]   rate: %d Hz\n", bank->len, newLen - 1, 0);

    // add new sample corresponding to this data block
    insertAfterLList(bank->samples, Sample_create(getSizeLList(bank->samples), bank->len, VGMCommand_getDataBlockLen(command), 0));

    // set new data and len
    bank->data = newData;
    bank->offset = 0;
    bank->len = newLen;
}

VGMCommand* SampleBank_getDataBlockCommand(SampleBank* bank)
{
    // return command
    return VGMCommand_createEx(bank->data, bank->offset, -1);
}

LList* SampleBank_getDeclarationCommands(SampleBank* bank)
{
    unsigned char* data;
    LList* result;

    // rebuild data
    data = malloc(5);
    data[0] = VGM_STREAM_CONTROL;
    data[1] = bank->id;
    data[2] = 0x02;
    data[3] = 0x00;
    data[4] = 0x2A;
    result = createElement(VGMCommand_createEx(data, 0, -1));

    data = malloc(5);
    data[0] = VGM_STREAM_DATA;
    data[1] = bank->id;
    data[2] = bank->id;
    data[3] = 0x01;
    data[4] = 0x00;
    result = insertAfterLList(result, VGMCommand_createEx(data, 0, -1));

    return getHeadLList(result);
}

//Sample* SampleBank_getSampleByOffsetAndLen(SampleBank* bank, int dataOffset, int len)
//{
//    LList* list;
//
//    list = bank->samples;
//    while (list != NULL)
//    {
//        Sample* sample = list->element;
//        int frameSize = Sample_getFrameSize(sample);
//
//        // allow a margin of 1 frame for both offset and size
//        if ((abs(sample->dataOffset - dataOffset) < frameSize) && (abs(len - sample->len) < frameSize))
//            return sample;
//
//        list = list->next;
//    }
//
//    return NULL;
//}

Sample* SampleBank_getSampleByOffset(SampleBank* bank, int dataOffset)
{
    LList* list;

    list = bank->samples;
    while (list != NULL)
    {
        Sample* sample = list->element;
        int frameSize = Sample_getFrameSize(sample);

        // allow a margin of 1 frame for offset
        if (abs(sample->dataOffset - dataOffset) < frameSize)
            return sample;

        list = list->next;
    }

    return NULL;
}

Sample* SampleBank_getSampleById(SampleBank* bank, int id)
{
    LList* list;

    list = bank->samples;
    while (list != NULL)
    {
        Sample* sample = list->element;

        if (sample->id == id)
            return sample;

        list = list->next;
    }

    return NULL;
}

Sample* SampleBank_addSample(SampleBank* bank, int dataOffset, int len, int rate)
{
//    TODO: maybe use a switch here
//    Sample* result = SampleBank_getSampleByOffsetAndLen(bank, dataOffset, len);
    Sample* result = SampleBank_getSampleByOffset(bank, dataOffset);

    // not found --> create new sample
    if (result == NULL)
    {
        if (verbose)
            printf("Sample added     [%6X-%6X]  len: %6X  rate: %d Hz\n", dataOffset, dataOffset + (len - 1), len, rate);

        result = Sample_create(getSizeLList(bank->samples), dataOffset, len, rate);
        insertAfterLList(bank->samples, result);
    }
    // confirmation of sample
    else if (result->rate == 0)
    {
        if (verbose)
            printf("Sample confirmed [%6X-%6X]  len: %6X --> %6X   rate: %d --> %d Hz\n", dataOffset, dataOffset + (len - 1), result->len, len, result->rate, rate);

        result->rate = rate;
        result->len = len;
    }
    else
    {
        // adjust sample length if needed
        if (result->len < len)
        {
            if (verbose)
                printf("Sample modified  [%6X-%6X]  len: %6X --> %6X\n", dataOffset, dataOffset + (len - 1), result->len, len);

            result->len = len;
        }

//        // get size of sample frame
//        int frameLen = Sample_getFrameSize(result);
//
//        // less than 1 frame in size difference --> assume same sample
//        if (abs(len - result->len) < frameLen)
//        {
//            // adjust sample length if needed
//            if (result->len < len)
//            {
//                if (verbose)
//                    printf("Sample modified  [%6X-%6X]  len: %6X --> %6X\n", dataOffset, dataOffset + (len - 1), result->len, len);
//
//                result->len = len;
//            }
//        }
//        // too large difference in size --> assume we have a new sample here
//        else
//        {
//            if (verbose)
//                printf("Sample added     [%6X-%6X]  len: %6X  rate: %d Hz\n", dataOffset, dataOffset + (len - 1), len, rate);
//
//            result = Sample_create(getSizeLList(bank->samples), dataOffset, len, rate);
//            insertAfterLList(bank->samples, result);
//        }
    }

    return result;
}


Sample* Sample_create(int id, int dataOffset, int len, int rate)
{
    Sample* result;

    result = malloc(sizeof(Sample));

    result->id = id;
    result->dataOffset = dataOffset;
    result->len = len;
    result->rate = rate;

    return result;
}

int Sample_getFrameSize(Sample* sample)
{
    if (sample->rate > 0)
        return sample->rate / 60;

    // consider 4Khz by default for safety
    return 4000 / 60;
}

void Sample_setRate(Sample* sample, int value)
{
    if (value != 0)
    {
        // set sample rate if needed
        if (sample->rate != value)
        {
            if (verbose)
                printf("Sample modified  [%6X-%6X]  rate: %d --> %d Hz\n", sample->dataOffset, sample->dataOffset + (sample->len - 1), sample->rate, value);

            sample->rate = value;
        }
    }
}

VGMCommand* Sample_getSetRateCommand(SampleBank* bank, Sample* sample, int value)
{
    unsigned char* data;

    // build command
    data = malloc(6);
    data[0] = VGM_STREAM_FREQUENCY;
    data[1] = bank->id;
    data[2] = (value >> 0) & 0xFF;
    data[3] = (value >> 8) & 0xFF;
    data[4] = 0x00;
    data[5] = 0x00;

    return VGMCommand_createEx(data, 0, -1);
}

VGMCommand* Sample_getStartLongCommandEx(SampleBank* bank, Sample* sample, int value)
{
    unsigned char* data;
    int adjLen = min(value, sample->len);

    // build command
    data = malloc(11);
    data[0] = VGM_STREAM_START_LONG;
    data[1] = bank->id;
    data[2] = (sample->dataOffset >> 0) & 0xFF;
    data[3] = (sample->dataOffset >> 8) & 0xFF;
    data[4] = (sample->dataOffset >> 16) & 0xFF;
    data[5] = 0x00;
    data[6] = 0x01;
    data[7] = (adjLen >> 0) & 0xFF;
    data[8] = (adjLen >> 8) & 0xFF;
    data[9] = (adjLen >> 16) & 0xFF;
    data[10] = 0x00;

    return VGMCommand_createEx(data, 0, -1);
}

VGMCommand* Sample_getStartLongCommand(SampleBank* bank, Sample* sample)
{
    return Sample_getStartLongCommandEx(bank, sample, sample->len);
}

VGMCommand* Sample_getStopCommand(SampleBank* bank, Sample* sample)
{
    unsigned char* data;

    // build command
    data = malloc(2);
    data[0] = VGM_STREAM_STOP;
    data[1] = bank->id;

    return VGMCommand_createEx(data, 0, -1);
}

File: tools/crates/xgmtool/src/sample_bank.rs
use crate::vgm_command::*;

/// Represents a PCM sample within a VGM data bank
#[derive(Clone, Debug)]
pub struct Sample {
    pub id: usize,
    pub data_offset: usize,
    pub len: usize,
    pub rate: u32,
}

impl Sample {
    pub fn new(id: usize, data_offset: usize, len: usize, rate: u32) -> Self {
        Sample { id, data_offset, len, rate }
    }

    pub fn get_frame_size(&self) -> usize {
        if self.rate > 0 { self.rate as usize / 60 } else { 4000 / 60 }
    }

    pub fn set_rate(&mut self, value: u32) {
        if value != 0 && self.rate != value {
            if crate::is_verbose() {
                println!("Sample modified  [{:06X}-{:06X}]  rate: {} --> {} Hz",
                    self.data_offset, self.data_offset + self.len - 1, self.rate, value);
            }
            self.rate = value;
        }
    }

    pub fn get_set_rate_command(&self, bank_id: u8, value: u32) -> VGMCommand {
        let data = vec![
            VGM_STREAM_FREQUENCY, bank_id,
            (value & 0xFF) as u8, ((value >> 8) & 0xFF) as u8, 0x00, 0x00,
        ];
        VGMCommand::from_owned(data, -1)
    }

    pub fn get_start_long_command_ex(&self, bank_id: u8, len: usize) -> VGMCommand {
        let adj_len = len.min(self.len);
        let data = vec![
            VGM_STREAM_START_LONG, bank_id,
            (self.data_offset & 0xFF) as u8, ((self.data_offset >> 8) & 0xFF) as u8,
            ((self.data_offset >> 16) & 0xFF) as u8, 0x00, 0x01,
            (adj_len & 0xFF) as u8, ((adj_len >> 8) & 0xFF) as u8,
            ((adj_len >> 16) & 0xFF) as u8, 0x00,
        ];
        VGMCommand::from_owned(data, -1)
    }

    pub fn get_start_long_command(&self, bank_id: u8) -> VGMCommand {
        self.get_start_long_command_ex(bank_id, self.len)
    }

    pub fn get_stop_command(&self, bank_id: u8) -> VGMCommand {
        VGMCommand::from_owned(vec![VGM_STREAM_STOP, bank_id], -1)
    }
}

/// A bank of PCM sample data from a VGM file
#[derive(Clone, Debug)]
pub struct SampleBank {
    pub samples: Vec<Sample>,
    pub data: Vec<u8>,
    pub len: usize,
    pub id: u8,
}

impl SampleBank {
    pub fn create(command: &VGMCommand) -> Option<Self> {
        if !command.is_data_block() {
            println!("Error: incorrect sample data declaration at {:06X} !", command.offset);
            return None;
        }

        let id = command.get_data_bank_id();
        let block_len = command.get_data_block_len() as usize;

        if crate::is_verbose() {
            println!("Initial bank sample added [{:06X}-{:06X}]   rate: 0 Hz", 0, block_len - 1);
        }

        // The data starts at offset 7 in the command (after 0x67 0x66 type len[4])
        let data = command.data.clone();
        let samples = vec![Sample::new(0, 0, block_len, 0)];

        Some(SampleBank { samples, data, len: block_len, id })
    }

    pub fn add_block(&mut self, command: &VGMCommand) {
        if !command.is_data_block() { return; }

        let block_len = command.get_data_block_len() as usize;
        let new_len = self.len + block_len;

        // Build new data: existing header+data + new block data
        let mut new_data = self.data[..self.len + 7].to_vec();
        new_data.extend_from_slice(&command.data[7..7 + block_len]);
        crate::set_u32(&mut new_data, 3, new_len as u32);

        if crate::is_verbose() {
            println!("Initial block sample added [{:06X}-{:06X}]   rate: 0 Hz", self.len, new_len - 1);
        }

        let sample_id = self.samples.len();
        self.samples.push(Sample::new(sample_id, self.len, block_len, 0));
        self.data = new_data;
        self.len = new_len;
    }

    pub fn get_data_block_command(&self) -> VGMCommand {
        VGMCommand::from_owned(self.data.clone(), -1)
    }

    pub fn get_declaration_commands(&self) -> Vec<VGMCommand> {
        vec![
            VGMCommand::from_owned(vec![VGM_STREAM_CONTROL, self.id, 0x02, 0x00, 0x2A], -1),
            VGMCommand::from_owned(vec![VGM_STREAM_DATA, self.id, self.id, 0x01, 0x00], -1),
        ]
    }

    pub fn get_sample_by_offset(&self, data_offset: usize) -> Option<&Sample> {
        self.samples.iter().find(|s| {
            let frame_size = s.get_frame_size();
            (s.data_offset as i64 - data_offset as i64).unsigned_abs() < frame_size as u64
        })
    }

    pub fn get_sample_by_offset_mut(&mut self, data_offset: usize) -> Option<&mut Sample> {
        self.samples.iter_mut().find(|s| {
            let frame_size = s.get_frame_size();
            (s.data_offset as i64 - data_offset as i64).unsigned_abs() < frame_size as u64
        })
    }

    pub fn get_sample_by_id(&self, id: usize) -> Option<&Sample> {
        self.samples.iter().find(|s| s.id == id)
    }

    pub fn add_sample(&mut self, data_offset: usize, len: usize, rate: u32) -> usize {
        // Try to find existing sample at this offset
        if let Some(idx) = self.samples.iter().position(|s| {
            let frame_size = s.get_frame_size();
            (s.data_offset as i64 - data_offset as i64).unsigned_abs() < frame_size as u64
        }) {
            let sample = &mut self.samples[idx];
            if sample.rate == 0 {
                if crate::is_verbose() {
                    println!("Sample confirmed [{:06X}-{:06X}]  len: {:06X} --> {:06X}   rate: {} --> {} Hz",
                        data_offset, data_offset + len - 1, sample.len, len, sample.rate, rate);
                }
                sample.rate = rate;
                sample.len = len;
            } else if sample.len < len {
                if crate::is_verbose() {
                    println!("Sample modified  [{:06X}-{:06X}]  len: {:06X} --> {:06X}",
                        data_offset, data_offset + len - 1, sample.len, len);
                }
                sample.len = len;
            }
            return idx;
        }

        // New sample
        if crate::is_verbose() {
            println!("Sample added     [{:06X}-{:06X}]  len: {:06X}  rate: {} Hz",
                data_offset, data_offset + len - 1, len, rate);
        }
        let id = self.samples.len();
        self.samples.push(Sample::new(id, data_offset, len, rate));
        id
    }
}
